Improve port handling in tests

v0d1ch opened the pull request:
## Why
Our network tests that need to use free ports are flaky. We want to improve the situation and have lesser amount of flaky tests.

## What
Use hspec hook to introduce a `TVar` that will hold a list of _occupied_ ports so that we don't have ports clashing between tests.


---

<!-- Tick off or strike-through / remove if not applicable -->
* [ ] CHANGELOG updated
* [ ] Documentation updated
* [x] Added and/or updated haddocks
* [x] No new TODOs introduced or explained herafter

v0d1ch committed:
Improve port handling in tests

v0d1ch committed:
Use hspec hook to have only one TVar

v0d1ch committed:
Refactor and re-use `withPort`

- Also remove redundant functions

v0d1ch committed:
Remove used port after a test and use parallel

v0d1ch committed:
Rename and try to not free used port in a list

ch1bo commented on hydra-node/test/Hydra/API/ServerSpec.hs:
Must: Not move the API.ServerSpec

It's testing the `API.Server` module and we try to keep the names in sync.

ch1bo commented on hydra-node/test/Hydra/Port/SpecHook.hs:
```diff
@@ -0,0 +1,11 @@
+-- | Hspec hook to re-use the same 'TVar' for holding the list of all used ports.
+-- This is needed for the tests that depend on having a free port.
+hook :: SpecWith (TVar IO [Int]) -> Spec
+hook = beforeAll (newTVarIO [])
```
Must: find a different way to re-use a port counting variable across tests

We cannot just force all tests to live in the same directory if they need a free port. What if there is another resource which needs to be managed to avoid test pollution?

One option (maybe the only one I know) is to use `unsafeDupablePerformIO` to give you a CAF of `TVar IO [Int]` in this case. Something like `globalUsedPorts = unsafeDupablePerformIO $ newTVarIO []`. See also https://hackage.haskell.org/package/global-variables-1.0.1.1/docs/Data-Global.html for a package which uses this approach. A well encapsulated, marked as "Test-only" (!) module for providing free ports should do the trick here.

ch1bo commented on hydra-node/test/Hydra/Port/ServerSpec.hs:
```diff
@@ -0,0 +1,339 @@
+spec =
+  describe "ServerSpec" $
+    parallel $ do
+      it "greets" $ \tvar ->
```
Must: not litter the code with anonymous variables

The TVar littering here and above in the signature is to be avoided at all costs. It makes the code harder to read and impossible to know (without context) what this `tvar` is for? Even the type is very anonymous.

ch1bo commented on hydra-test-utils/src/Test/Network/Ports.hs:
```diff
@@ -40,61 +21,35 @@ import System.Random.Shuffle (
--- conditions. Production code should use better methods such as passing a
--- listening socket to the child process.
-randomUnusedTCPPorts :: Int -> IO [Int]
-randomUnusedTCPPorts count = do
```
Isn't this the source of all race conditions? The documentation of `openFreePort` and especially on [getFreePort](https://hackage.haskell.org/package/port-utils-0.2.1.0/docs/Network-Socket-Free.html#v:getFreePort) (a factored out helper from `warp` we should use`) as well as our code states that the race condition is only when other processes do bind that port between `getFreePort` and binding to it again.

HOWEVER, this code here seemingly is still used (as you updated it), does also tests ports as open .. on a shuffled, self-made list! This basically circumvents the mechanism used in `openFreePort`/`getFreePort`, which asks the OS to give a free port. The OS does not know of the list we created in this function!

Furthermore, using a `TVar` only safe guards within a process (if we don't use this other mechanism).. the OS mechanism should work system-wide.

ch1bo reviewed (changes requested):
Two things:
* The solution idea should not litter and pollute the whole code. Using `unsafeDupablePerformIO` would be an (IMO) acceptable trade-off versus the noise.

* I believe just dropping the usage of `randomUnusedTCPPorts` should fix our problems.
  - We should use [getFreePort](https://hackage.haskell.org/package/port-utils-0.2.1.0/docs/Network-Socket-Free.html#v:getFreePort) from this non-warp package though as it has less dependencies.

v0d1ch committed:
Go back and just reuse `withFreePort`

v0d1ch commented:
@ch1bo seems like `TVar` is not needed in the end. Thanks for the `Data.Global` mention, I wasn't aware this existed.

ch1bo reviewed (approved):
Nice. Hope this helps now :)

ch1bo commented:
Unfortunately tests are red in this version. The logs don't say _exactly_ why they fail (I would expect socket bind errors or so). Also, I just realized that we are running property tests against these real server/client interactions. That is, we use 100 random free ports per test case. Which quickly sums up a lot and maybe this approach does not scale as much as the property tests need? i.e. it's not about re-using the port, but something goes wrong when acquiring it now via the system-level `getFreePort`?

We should find out why the `withClient` times out: did it receive any messages? did the server crash / not come up?

v0d1ch committed:
Use a proper tracer

v0d1ch committed:
Introduce a test to assert we can't run the server on the same port

ch1bo committed:
Wait befor calling action in withAPIServer

This avoids race conditions in code which wants to create a connection
to the server right away. See also https://hackage.haskell.org/package/warp-3.3.25/docs/Network-Wai-Handler-Warp.html#v:setBeforeMainLoop

ch1bo committed:
Not retry on failed client connection

This should not be necessary anymore as the server only calls it's
continuation when it can accept connections. So we see problems right
way and not just "timeout after 5".

ch1bo committed:
Make a APIServer test not a property

Running the test with transaction output configurability does not
benefit from doing it different transactions (which are just integers in
SimpleTx anyway)

ch1bo committed:
Move APIServer trace and this reduces flakyness

This is obviously smelly and we left a comment.
